crates/edr_provider/src/data.rs

```diff
@@ -1414,16 +1419,19 @@ impl<LoggerErrorT: Debug> ProviderData<LoggerErrorT> {
             },
         )?;
 
-        let block_number = self.blockchain.last_block_number();
-        let state_root = self.state.state_root()?;
+        let state_root = state.state_root()?;
 
+        self.mem_pool.update(&state)?;
+
+        let block_number = self.blockchain.last_block_number();
         self.irregular_state
             .state_override_at_block_number(block_number)
             .or_insert_with(|| StateOverride::with_state_root(state_root))
             .diff
             .apply_account_change(address, account_info.clone());
 
-        self.mem_pool.update(&self.state)?;
+        self.block_state_cache.push(block_number, state.clone());
+        self.state = state;
```

Comment:
This ensures that the state is not modified when there are previous failures in `fn state_root` or `fn update` calls.

This occurs at the cost of an extra clone. Once we adopt an immutable Merkle-Patricia trie, this cost should become negligible.